### DISCUSSION TITLE
📎 Improve `useNamingConvention` options

### DISCUSSION THREAD

**Discussion by anonrig** [biomejs/biome#1725] [Ideas]
### Description

It would be nice to support all cases of `@typescript-eslint/naming-convention` rule, such as the following usage:

```
    '@typescript-eslint/naming-convention': [
      'error',
      {
        selector: 'typeLike',
        format: ['PascalCase'],
        leadingUnderscore: 'allow',
      },
      {
        selector: 'enumMember',
        format: ['UPPER_CASE'],
      },
    ],
```

**Comment by Conaclos** [comment]
Note that we already have an option for enum member case. I am open to relax some default and adding some options. Could you be more explicit about the need?

**Reply by nouman2075** [reply]
Using `useNamingConvention` only allow variables in camelCase. Is there's a possiblity to set different casing for variable names like snake_case or any other case? Same for fields in  type and interface.

Use case:
We are using rust for backend and all the fields are in snake_case. We want to have functions in camelCase but variables, fields in type and interface in snake_case

**Reply by anonrig** [reply]
@Conaclos Basically, this section of the example config I've shared is not supported by Biome:

```
{
        selector: 'typeLike',
        format: ['PascalCase'],
        leadingUnderscore: 'allow',
      },
```

**Reply by Conaclos** [reply]
@anonrig The rule already enforces PascalCase for every type-like entity and allow leading underscore. Have you a need to enforce other cases for type-like entities?

**Reply by Conaclos** [reply]
> We are using rust for backend and all the fields are in snake_case. We want to have functions in camelCase but variables, fields in type and interface in snake_case

@nouman2075 I understand your use case. That's really specific. Let me think about that. By the way, why do you need variables in snake_case?

**Reply by nouman2075** [reply]
We want to enforce rule that variables, fields in type and interface should follow same casing. If we stick to camelCase then if we try to pass snake_case fields received from backend to other components, biome will throw error. Also the code will be bad if we use camelCase and snake_case together. So we should stick to one casing.

Just like Biome provide option for enum member, it should provide separate option for variables, functions, typeLike...
Here's the list of possible cases for [naming convection in eslint](https://typescript-eslint.io/rules/naming-convention/)

**Reply by Conaclos** [reply]
Thanks for these details. I will propose a change to support your use cases.

**Reply by ValentinGurkov** [reply]
Mocking also seems a bit difficult since if I attempt to mock an exported class in PascalCase, it forces me to use camelCase:
![image](https://github.com/biomejs/biome/assets/35373940/38ac4270-c023-45fc-a4d6-1a90c56e6cf8)
But then it won't match the export.

**Comment by Conaclos** [comment]
## Proposal v1

I have spent some time thinking about your needs and more generally the needs of the community.

I propose to add a way to select a given set of names to apply a custom format and require a prefix/suffix. Unlike the ESLint rule, we provide a limited set of selectors.

Some selectors override others. For example, `topLevelConst` overrides `const`; `const` overrides `variables`. If a property is not set, the selector inherits from the property of the overridden selector.

```ts
type Case =  "camelCase" |  "PascalCase" | "CONSTANT_CASE" | "snake_case" 

type NamingConventionOptions = {
  strictCase?: boolean, // same option as now
  selectors: {
    [selector: Selector]?: {
      format?: Case[],
      prefix?: string[],
      suffix?: string[],
      exceptions?: string[],
    },
  },
  requireAscii: boolean,
  ignoreQuoted: boolean,
  /** @Deprecated */ enumMemberCase: Case,
}

type Selector  =
  | "default"
  | "typeLike" // class, interface, type alias, type parameter, enum
  | "interface"
  | "function" // function declaration, function expression
  | "namespaceLike" // namespace, import namespace, export namespace
  | "enumMember"
  | "method"
  | "property" // class prop, prop parameter, type prop, object prop
  | "privateProperty" // property with `private` or `protected` accessibility modifier
  | "objectProperty"
  | "staticProperty"
  | "staticReadonlyProperty"
  | "parameter"
  | "variable"
  | "const"
  | "topLevelConst"
```

Some examples of selectors:

- Require a leading underscore for private and protected properties
  ```js
  {
    privateProperty: { prefix: ["_"] }
  }
  ```

- Allow a leading underscore for private and protected properties
  ```js
  {
    privateProperty: { prefix: ["_", ""] } // order is important
  }
  ```

- Allow snake_case for types
  ```js
  {
    typeLike: { format: ["PascalCase", "snake_case"] } // order is important
  }
  ```

- Require interface with a `I` prefix, except for `MyException`
  ```js
  {
    interface: { prefix: ["I"], exceptions: ["MyException"] } // order is important
  }
  ```

With these new options, we could forbid leading and trailing underscores by default. The only exceptions are unused variables, which can always start with a single underscore.

We still accept a leading and trailing dollar sign. This can be overridden in the `default' selector.

**Reply by PointSingularity** [reply]
This looks awesome! 
If this was implemented, it would allow me to finally fully migrate my work's legacy ESLint config:
```ts
  '@typescript-eslint/naming-convention': [
      'error',
      {
        selector: 'interface',
        format: ['PascalCase'],
        leadingUnderscore: 'forbid',
      },
      {
        selector: 'variable',
        format: ['camelCase', 'UPPER_CASE', 'snake_case'],
        leadingUnderscore: 'allow',
        custom: {
          regex: '^(any|Number|number|String|string|Boolean|boolean|Undefined|undefined)$',
          match: false,
        },
      },
    ],
```

**Reply by rujorgensen** [reply]
This proposal looks very interesting to me as well. I'm looking into potentially switching from eslint to Biome at work because of the excellent performance and helpful error descriptions. We have conventions for type/enum/interface -names where we sometimes prefix with "T", "E", or "I", and suffix with "_S". Can this be tracked somewhere?

**Reply by Conaclos** [reply]
Here #1900

**Comment by hilja** [comment]
Would it make sense to hard-code some exceptions for selected industry standards like S3? Or does it go against the Biome’s philosophy?

For some reason the S3 SDK does this:

```js
S3.PutObject({
  Bucket: S3_IMAGES_BUCKET,
  Key,
  Body,
  ContentEncoding: 'base64',
  ContentType: img.mime,
  CacheControl: 'public, max-age=31536000', // year
})
```

**Reply by Conaclos** [reply]
I want to keep the rule general. I plan to add some options to configure the preferences. This should make possible to configure in a way that satisfy your needs.

However, it is also really important for us to have sensible defaults. I am wondering more and more if we should not accept `snake_case` and `PascalCase` by default for literal object properties.

**Reply by hilja** [reply]
Will Biome have the concept of environments? One option could be to offer an S3 or whatever env for popular quirky packages.

> I am wondering more and more if we should not accept snake_case and PascalCase by default for literal object properties.

But you aren't accepting. The above code will produce a lint error.

I'd be okay to have a file-level ignore comment to disable PascalCase per file.

**Reply by Conaclos** [reply]
> But you aren't accepting. The above code will produce a lint error.

Maybe my explanation is confusing: What I mean is allowing by default `camelcase`, `snake_case` and `PascalCase` for literal object properties. For now, we only accept `camelcase`.

**Reply by Conaclos** [reply]
> Will Biome have the concept of environments? One option could be to offer an S3 or whatever env for popular quirky packages.

It is not decided yet. However, I don't think so. Have you some references/docs about S3 conventions.

**Reply by hilja** [reply]
> allowing by default camelcase, snake_case and PascalCase for literal object properties

Ah, that makes sense :)

[Here's the input docs for `PutObject`](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/Package/-aws-sdk-client-s3/Interface/PutObjectCommandInput/), all properties are capitalized for some reason, and some use uppercase acronyms `SSEKMSEncryptionContext` etc.

**Reply by hilja** [reply]
> if we should not accept snake_case and PascalCase by default for literal object properties.

This is actually a great idea, now that I've been aware of this more there are a _lot_ of APIs that use snake_case or PascalCase. For example the Stripe API and most Postgres related things, because in pg case sensitive strings needs to be quoted: `SELECT * FROM "MyPosts"` so it's just easier to use snake_case.

**Reply by arendjr** [reply]
I'm not sure how I would feel about making the default more lenient for specific APIs. Personally, I would prefer to keep usage of such APIs constrained to specific modules/services, so I would configure `overrides` that allow specific patterns in those files or folders. I wouldn't like it if those patterns were allowed throughout a codebase though.

**Comment by Conaclos** [comment]
I've just published #2770 that adds options to `useNamingConvention`. Any opinions/reviews is welcomed :)

**Comment by muriiloandrade** [comment]
Hey guys, beforehand I apoligize if that's not the right place for this, but can you explain how can we enforce naming conventions on abstract classes?
I'm failing to configure it and the obvious way for me would be like this:
```json
...
"conventions": [
  {
    "selector": {
      "kind": "class",
      "modifiers": ["abstract"]
    },
    "formats": ["PascalCase"]
  }
]
...
```
My IDE is throwing `The 'abstract' modifier cannot be used with the 'class' kind. biome(deserialize)` at my face and I don't know if it's a bug, if it's a missing option or it's intended to not allow it. Please help 🫠

**Reply by Conaclos** [reply]
This is a bug. I will take a look. Unfortunately, I think that there is no workaround...

**Reply by Conaclos** [reply]
Fixed in #3473. We will have to wait for the 1.9 release.

**Reply by muriiloandrade** [reply]
Thanks a lot for the quick answer and fix @Conaclos! I'll be testing it as soon as it releases 🤘🚀

**Comment by ZettZet** [comment]
Is there a way to migrate that rule? There is no 'exported' selector

```js
'@typescript-eslint/naming-convention': [
  'error',
  {
    selector: ['variable'],
    suffix: ['Api'],
    format: ['camelCase'],
    modifiers: ['exported', 'const'],
  },
  {
    selector: ['function'],
    suffix: ['Api'],
    format: ['camelCase'],
    modifiers: ['exported'],
  },
]
```

**Reply by Conaclos** [reply]
We are not currently able to select exported variable/functions. However, you can select `const`/`function` in the global scope. This should cover most of the cases and unfortunately this will include variables and functions in global scope that are not exported.

This could lead to something like:

```json5
"conventions": [
  {
    "selector": {
      "kind": "const",
      "scope": "global"
    },
    "match": "(.*)Api",
    "formats": ["camelCase"]
  },
  {
    "selector": {
      "kind": "function",
      "scope": "global"
    },
    "match": "(.*)Api",
    "formats": ["camelCase"]
  }
]
```

We could evaluate the addition of a query for exported entities because this seems to be a reasonable addition.
